src/delay/timer.rs

```diff
@@ -0,0 +1,193 @@
+                        bb::set(&rcc.$apbrstr, $en_bit);
+                        bb::clear(&rcc.$apbrstr, $en_bit);
+                    }
+
+                    // Enable one-pulse mode (counter stops counting at the next update
+                    // event, clearing the CEN bit)
+                    tim.cr1.modify(|_, w| w.opm().enabled());
+
+                    Self { tim, clocks }
+                }
+
+                /// Releases the timer resource
+                pub fn free(self) -> pac::$TIM {
+                    self.tim
+                }
+            }
+
+            impl DelayUs<u32> for $struct {
+                /// Sleep for up to 2^32-1 microseconds (~71 minutes).
+                fn delay_us(&mut self, us: u32) {
+                    // Set up prescaler so that a tick takes exactly 1 µs.
+                    //
+                    // For example, if the clock is set to 48 MHz, with a prescaler of 48
+                    // we'll get ticks that are 1 µs long. This means that we can write the
+                    // delay value directly to the auto-reload register (ARR).
+                    let psc = u16(self.clocks.pclk1().0 / 1_000_000)
+                        .expect("Prescaler does not fit in u16");
+                    let arr = us;
+                    $waitfn(&mut self.tim, psc, arr);
+                }
+            }
+
+            impl DelayUs<u16> for $struct {
+                /// Sleep for up to 2^16-1 microseconds (~65 milliseconds).
+                fn delay_us(&mut self, us: u16) {
+                    // See DelayUs<u32> for explanations.
+                    let psc = u16(self.clocks.pclk1().0 / 1_000_000)
+                        .expect("Prescaler does not fit in u16");
+                    let arr = u32(us);
+                    $waitfn(&mut self.tim, psc, arr);
+                }
+            }
+
+            impl DelayMs<u32> for $struct {
+                /// Sleep for up to (2^32)/2-1 milliseconds (~24 days).
+                /// If the `ms` value is larger than 2147483647, the code will panic.
+                fn delay_ms(&mut self, ms: u32) {
+                    // See next section for explanation why the usable range is reduced.
+                    assert!(ms <= 2_147_483_647); // (2^32)/2-1
+
+                    // Set up prescaler so that a tick takes exactly 0.5 ms.
+                    //
+                    // For example, if the clock is set to 48 MHz, with a prescaler of 24'000
+                    // we'll get ticks that are 0.5 ms long. This means that we can write the
+                    // delay value multipled by two to the auto-reload register (ARR).
+                    //
+                    // Note that we cannot simply use a prescaler value where the tick corresponds
+                    // to 1 ms, because then a clock of 100 MHz would correspond to a prescaler
+                    // value of 100'000, which doesn't fit in the 16-bit PSC register.
```

Comment:
I'm not 100% sure if APB1 follows the system clock, but I think there are F4 chips that support 100 MHz or more.

Reply:
Uh, actually there are models (e.g. STM32F469) that support system clocks up to 180 MHz. In that case, the even with a prescaler for 0.5 ms, the value would be 90'000, which doesn't fit in 16 bit.

Any ideas what to do? Should we use a prescaler for 0.25 ms? That would reduce the usable range to (2^32)/4, which is still 12 days for 32 bit values. I think that should be plenty.

Alternatively we could dynamically calculate the PSC and ARR, at the cost of either the range or the resolution. I'd usually prefer resolution over range for a generic implementation. However, a dynamic implementation that might panic for certain values depending on the system clock configuration sucks.